### PR TITLE
feat: add redeem printing v2 action and update instant sale

### PR DESCRIPTION
**Pull request by kurpav**
Add redeem printing v2 action and updates instant sale to support the winning type

Tested on metaplex-foundation repo

### CODE REVIEW

File: src/Account.ts [hunk 1]
```diff
@@ -38,7 +39,7 @@ export class Account<T = unknown> {
   static async getInfo(connection: Connection, pubkey: AnyPublicKey) {
     const info = await connection.getAccountInfo(new PublicKey(pubkey));
     if (!info) {
-      throw new Error(`Unable to find account: ${pubkey}`);
+      throw ERROR_ACCOUNT_NOT_FOUND(pubkey);
```
Comment (b2kdaman) [inline src/Account.ts:42]: Method shouldn't be cased like this

File: src/actions/instantSale.ts [hunk 2]
```diff
@@ -31,89 +28,59 @@ export const instantSale = async ({
...
-    metadata,
-  });
+  // NOTE: it's divided into 3 transactions since transaction size is restricted
+  if (winningConfigType === WinningConfigType.FullRightsTransfer) {
+    const { txId } = await redeemFullRightsTransferBid({ connection, wallet, store, auction });
+    txIds.push(txId);
+  } else if (winningConfigType === WinningConfigType.PrintingV2) {
+    const { txId } = await redeemPrintingV2Bid({ connection, wallet, store, auction });
+    txIds.push(txId);
+  }
```
Comment (b2kdaman) [inline src/actions/instantSale.ts:74]: I'd rather use switch here

File: src/actions/redeemFullRightsTransferBid.ts [hunk 3]
```diff
@@ -103,7 +103,7 @@ interface IRedeemBidTransactionsParams {
   metadata: PublicKey;
 }
 
-export const getRedeemBidTransactions = async ({
+export const getRedeemFullRightsTransferBidTransactions = async ({
```
Comment (b2kdaman) [inline src/actions/redeemFullRightsTransferBid.ts:106]: Seven words...

File: src/actions/redeemPrintingV2Bid.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+  const newEditionPDA = await Edition.getPDA(newMint);
+
+  const metadataPDA = await Metadata.getPDA(originalMint);
+  const masterEditionPDA = await MasterEdition.getPDA(originalMint);
+  const masterEdition = await MasterEdition.load(connection, masterEditionPDA);
+
+  const prizeTrackingTicketPDA = await PrizeTrackingTicket.getPDA(auctionManagerPDA, originalMint);
+
+  let prizeTrackingTicket: PrizeTrackingTicket;
+  // first time we do redeem this account doesn't exist yet
```
Comment (b2kdaman) [inline src/actions/redeemPrintingV2Bid.ts:73]: Please update wording

File: src/actions/redeemPrintingV2Bid.ts [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  metadata: PublicKey;
+  newMint: PublicKey;
+  newMetadata: PublicKey;
+  newEdition: PublicKey;
+  masterEdition: PublicKey;
+  editionMarker: PublicKey;
+  prizeTrackingTicket: PublicKey;
+  winIndex: BN;
+  editionOffset: BN;
+}
```
Comment (b2kdaman) [inline src/actions/redeemPrintingV2Bid.ts:161]: Move interface to the top

File: src/actions/redeemPrintingV2Bid.ts [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+  txBatch.addTransaction(updatePrimarySaleHappenedViaTokenTx);
+  ////
+
+  return txBatch;
+};
+
+export function getEditionOffset(winIndex: number) {
+  const offset = new BN(1);
+  // NOTE: not sure if this the right way to calculate it
+  return offset.add(new BN(winIndex));
```
Comment (b2kdaman) [inline src/actions/redeemPrintingV2Bid.ts:239]: @austbot thoughts?

File: src/errors.ts [hunk 7]
```diff
@@ -24,3 +27,9 @@ export const ERROR_INVALID_ACCOUNT_DATA: () => MetaplexError = () => {
 export const ERROR_DEPRECATED_ACCOUNT_DATA: () => MetaplexError = () => {
   return new MetaplexError(ErrorCode.ERROR_DEPRECATED_ACCOUNT_DATA, 'Account data is deprecated');
 };
+
+export const ERROR_ACCOUNT_NOT_FOUND: (pubkey: AnyPublicKey) => MetaplexError = (
```
Comment (b2kdaman) [inline src/errors.ts:31]: We need to refactor this

**Review by b2kdaman — CHANGES_REQUESTED** [review]

File: src/programs/auction/accounts/Auction.ts [hunk 8]
```diff
@@ -73,14 +73,18 @@ export class BidState extends Borsh.Data<BidStateArgs> {
...
+      return null;
+    }
 
     const index = this.bids.findIndex((b) => b.key === bidder);
     // auction stores data in reverse order
     if (index !== -1) {
       const zeroBased = this.bids.length - index - 1;
       return zeroBased < this.max.toNumber() ? zeroBased : null;
-    } else return null;
+    } else {
```
Comment (zaxozhu) [inline src/programs/auction/accounts/Auction.ts:85]:
```suggestion
```

File: src/programs/auction/accounts/Auction.ts [hunk 9]
```diff
@@ -73,14 +73,18 @@ export class BidState extends Borsh.Data<BidStateArgs> {
...
+      return null;
+    }
```
Comment (zaxozhu) [inline src/programs/auction/accounts/Auction.ts:87]:
```suggestion

```

**Review by zaxozhu — APPROVED** [review]

File: src/errors.ts (hunk 7 above)
Comment (kurpav) [reply src/errors.ts:31]: Will be done in a separate PR

File: src/actions/redeemPrintingV2Bid.ts [hunk 10]
```diff
@@ -70,7 +95,7 @@ export const redeemPrintingV2Bid = async ({
   const prizeTrackingTicketPDA = await PrizeTrackingTicket.getPDA(auctionManagerPDA, originalMint);
 
   let prizeTrackingTicket: PrizeTrackingTicket;
-  // first time we do redeem this account doesn't exist yet
+  // this account doesn't exist when we do redeem for the first time
```
Comment (b2kdaman) [inline src/actions/redeemPrintingV2Bid.ts:98]: 🥇

**Review by b2kdaman — APPROVED** [review]

**Pull request merged**